File: src/2025-02-27-deepclaude37-rust/javascript/exercises/practice/two-bucket/two-bucket.js
//
// This is only a SKELETON file for the 'Two Bucket' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class TwoBucket {
  constructor(bucketOne, bucketTwo, target, startBucket) {
    // Check if target is larger than both buckets
    if (target > bucketOne && target > bucketTwo) {
      throw new Error('Target cannot be larger than both buckets');
    }
    
    // Check if target is achievable using GCD
    const gcd = (a, b) => {
      while (b !== 0) {
        let temp = b;
        b = a % b;
        a = temp;
      }
      return a;
    };
    
    const gcdValue = gcd(bucketOne, bucketTwo);
    if (target % gcdValue !== 0) {
      throw new Error('Target is not reachable');
    }

    this.bucketOneCap = bucketOne;
    this.bucketTwoCap = bucketTwo;
    this.target = target;
    this.startBucket = startBucket;
  }

  solve() {
    const queue = [];
    const visited = new Set();

    // Set initial state after filling the starting bucket (first action)
    let initialA, initialB;
    if (this.startBucket === 'one') {
      initialA = this.bucketOneCap;
      initialB = 0;
    } else {
      initialA = 0;
      initialB = this.bucketTwoCap;
    }

    queue.push({ a: initialA, b: initialB, moves: 1 });
    visited.add(`${initialA},${initialB}`);

    while (queue.length > 0) {
      const current = queue.shift();

      // Check if target has been reached
      if (current.a === this.target) {
        return {
          moves: current.moves,
          goalBucket: 'one',
          otherBucket: current.b
        };
      }
      if (current.b === this.target) {
        return {
          moves: current.moves,
          goalBucket: 'two',
          otherBucket: current.a
        };
      }

      // Generate all possible next states
      const nextStates = [
        { a: this.bucketOneCap, b: current.b, moves: current.moves + 1 }, // Fill bucket one
        { a: current.a, b: this.bucketTwoCap, moves: current.moves + 1 }, // Fill bucket two
        { a: 0, b: current.b, moves: current.moves + 1 }, // Empty bucket one
        { a: current.a, b: 0, moves: current.moves + 1 }, // Empty bucket two
        { // Pour from bucket one to bucket two
          a: current.a - Math.min(current.a, this.bucketTwoCap - current.b),
          b: current.b + Math.min(current.a, this.bucketTwoCap - current.b),
          moves: current.moves + 1
        },
        { // Pour from bucket two to bucket one
          a: current.a + Math.min(current.b, this.bucketOneCap - current.a),
          b: current.b - Math.min(current.b, this.bucketOneCap - current.a),
          moves: current.moves + 1
        }
      ];

      for (const state of nextStates) {
        // Check if the state violates the rule: starting bucket empty and other bucket full
        const isInvalid = this.startBucket === 'one' 
          ? (state.a === 0 && state.b === this.bucketTwoCap)
          : (state.b === 0 && state.a === this.bucketOneCap);

        if (!isInvalid) {
          const key = `${state.a},${state.b}`;
          if (!visited.has(key)) {
            visited.add(key);
            queue.push(state);
          }
        }
      }
    }

    throw new Error('No solution exists');
  }
}

File: src/2025-02-27-deepclaude37-rust/go/exercises/practice/two-bucket/two_bucket.go
package twobucket

import (
	"errors"
)

func Solve(sizeBucketOne, sizeBucketTwo, goalAmount int, startBucket string) (string, int, int, error) {
	// Validate parameters
	if sizeBucketOne <= 0 || sizeBucketTwo <= 0 {
		return "", 0, 0, errors.New("invalid bucket sizes")
	}
	if goalAmount <= 0 {
		return "", 0, 0, errors.New("invalid goal amount")
	}
	if startBucket != "one" && startBucket != "two" {
		return "", 0, 0, errors.New("invalid start bucket")
	}
	if goalAmount > sizeBucketOne && goalAmount > sizeBucketTwo {
		return "", 0, 0, errors.New("goal is larger than both buckets")
	}

	// Check initial state
	var initialB1, initialB2 int
	if startBucket == "one" {
		initialB1 = sizeBucketOne
		initialB2 = 0
		if initialB1 == goalAmount {
			return "one", 1, initialB2, nil
		}
	} else {
		initialB1 = 0
		initialB2 = sizeBucketTwo
		if initialB2 == goalAmount {
			return "two", 1, initialB1, nil
		}
	}

	// Proceed with BFS
	type state struct {
		b1, b2 int
	}

	initialSteps := 1
	initialState := state{initialB1, initialB2}

	queue := []struct {
		s     state
		steps int
	}{
		{initialState, initialSteps},
	}

	visited := make(map[state]bool)
	visited[initialState] = true

	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]

		currentState := current.s
		currentSteps := current.steps

		// Check if current state is a solution
		if currentState.b1 == goalAmount {
			return "one", currentSteps, currentState.b2, nil
		}
		if currentState.b2 == goalAmount {
			return "two", currentSteps, currentState.b1, nil
		}

		// Generate all possible next states
		nextStates := []state{
			{sizeBucketOne, currentState.b2}, // Fill bucket one
			{currentState.b1, sizeBucketTwo}, // Fill bucket two
			{0, currentState.b2},             // Empty bucket one
			{currentState.b1, 0},             // Empty bucket two
			func() state { // Pour from one to two
				amount := min(currentState.b1, sizeBucketTwo-currentState.b2)
				return state{currentState.b1 - amount, currentState.b2 + amount}
			}(),
			func() state { // Pour from two to one
				amount := min(currentState.b2, sizeBucketOne-currentState.b1)
				return state{currentState.b1 + amount, currentState.b2 - amount}
			}(),
		}

		for _, ns := range nextStates {
			// Check if the new state is invalid (starting bucket empty and other full)
			if (startBucket == "one" && ns.b1 == 0 && ns.b2 == sizeBucketTwo) ||
				(startBucket == "two" && ns.b2 == 0 && ns.b1 == sizeBucketOne) {
				continue
			}

			// Skip invalid states or already visited states
			if ns.b1 < 0 || ns.b1 > sizeBucketOne || ns.b2 < 0 || ns.b2 > sizeBucketTwo || visited[ns] {
				continue
			}

			visited[ns] = true
			queue = append(queue, struct {
				s     state
				steps int
			}{ns, currentSteps + 1})
		}
	}

	// No solution found
	return "", 0, 0, errors.New("no solution possible")
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
